### DISCUSSION TITLE
Is it possible to remove a notification?

### DISCUSSION THREAD

**Discussion by romerod** [andrei-tatar/node-red-contrib-smartnora#114] [Q&A]
If a notification was sent with a tag I understand that it can be replaced by a new notification, but it it also possible to remove it?

Example use case:
- Notification when any window is opened
- Removed when all windows are closed

**Comment by andrei-tatar** [comment]
It's not possible yet but it is technically possible to do it. 
How do you suggest the payload should look like for closing a notification?

**Reply by romerod** [reply]
I wanted to take a look in the backend code, to see if I can give a better answer by looking at the current implementation, but it's not opensource, right?

**Reply by andrei-tatar** [reply]
@romerod no, it's not open source yet. Maybe at some point in the future...

**Reply by romerod** [reply]
I was searching for the documentation of the google API and as I've understood notifications are tied to a device/trait, what trait are you using for notifications?

Was trying to find out if a image can be sent in the content and the possible formats of the content.

**Reply by andrei-tatar** [reply]
The notifications that are sent right now from NORA are web push notifications, not Google Home. I haven't implemented yet the ability to send device notifications.

https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification

You can send in the payload `icon` or `image` properties - a USVString containing the URL of an image to be used as an icon by the notification..

**Reply by romerod** [reply]
ah thanks

**Comment by romerod** [comment]
What about a clear checkbox which would disable Title, Body, Icon?

**Reply by normanr** [reply]
and could be overridden by `payload.close = true`?

**Reply by andrei-tatar** [reply]
sounds good to me

**Reply by andrei-tatar** [reply]
@romerod @normanr 

Fun fact: it's not possible to hide a notification without showing another one (https://stackoverflow.com/questions/33092065/google-chrome-silent-push-notifications).

So hiding works but the problem is that the browser generates a default one (`This site has been updated in the background`) if a notification is not shown. 

Was thinking of showing a generic: `One or more notifications were removed`.
What do you guys think?

**Reply by normanr** [reply]
That question (and the article it links to) is quite old. The docs at https://developer.mozilla.org/en-US/docs/Web/API/Push_API indicate that you don't have to show a notification, but some browsers may limit the number of pushes allowed if they do not end up showing a notification. Chrome doesn't seem to have a limit.

**Reply by andrei-tatar** [reply]
I tested with latest Chrome and I got the default notification pretty soon `This site has been updated in the background`. Might be the notification is removed when adding the PWA to the main screen. Will have to check it out.

**Reply by normanr** [reply]
Ahh, I forgot to write the second half of my comment. I noticed that there's a `ServiceWorkerRegistration.getNotifications()` method which allows you to retrieve a list of `Notification` objects. A `Notification` has a `close()` method, which I think is what should be used instead of sending another notification to replace the previous one.

**Reply by andrei-tatar** [reply]
@normanr Yup, that's exactly what I'm doing. The problem is that `Chrome` doesn't allow silent notifications. After a small number of attempts, it will show *This site has been updated in the background* notification. I can push the `silent` code to the wild if you want to play with it.

![image](https://user-images.githubusercontent.com/6473183/147506303-a59b6319-ea39-48bd-8cfe-3774cb90f1cb.png)

**Reply by normanr** [reply]
Why do you need to call `showNotification()` again? Why not just call `found?.close();` and leave it at that?

**Reply by andrei-tatar** [reply]
@normanr I pushed the code below for now. Make sure you update the app. Let me know how it works for you.
![image](https://user-images.githubusercontent.com/6473183/147545098-97a98957-7efd-497d-9b5c-4e7d14627c02.png)

**Reply by andrei-tatar** [reply]
@normanr Did you manage to have a play with it? Does it work for your browser?

**Reply by normanr** [reply]
I tested it with Chrome 97 and it works as expected (can close notifications without an additional message showing).

The code for Chrome is at https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/push_messaging/push_messaging_notification_manager.cc. The PushUserVisibleStatus stats van be seen at chrome://histograms/#PushMessaging.UserVisibleStatus. 

Interestingly I managed to test several silent closes, until I pushed too many silent closes in a row. I think I ran over budget at that point, and now all my closes generate the default message. (It also matters if the Smart NORA site is open or not, if it's open then budget is not considered.)

**Comment by romerod** [comment]
Might this depend on the Browser? For my example use case I'd prefer keeping the "all closed" instead

Holen Sie sich Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by andrei-tatar** [reply]
It is dependent on the browser, but I don't think there is a browser that allows silent push messages. 
Released support for closing notifications in version `1.6.0`. Awaiting feedback
